source/text.c

```diff
@@ -249,24 +283,242 @@ void C2D_DrawText(const C2D_Text* text, u32 flags, float x, float y, float z, fl
 	}
 	if (flags & C2D_WithColor)
 		color = va_arg(va, u32);
+	if (flags & C2D_WordWrap)
+		maxWidth = (float)va_arg(va, double); // Passed as float, but varargs promotes to double.
 
 	va_end(va);
 
 	C2Di_SetCircle(false);
 
-	for (cur = begin; cur != end; ++cur)
+	// Get information about lines
+	struct line {
+		u32 words;
+		u32 wordStart;
+	};
+	struct line lines[text->lines];
+	memset(lines, 0, sizeof(struct line) * text->lines);
+	for (cur = begin; cur != end; cur++)
+	{
+		if (cur->wordNo >= lines[cur->lineNo].words)
+			lines[cur->lineNo].words = cur->wordNo + 1;
+	}
+	for (u32 i = 1; i < text->lines; i++)
+	{
+		lines[i].wordStart = lines[i-1].wordStart + lines[i-1].words;
+	}
+
+	// Get information about words
+	struct word {
+		C2Di_Glyph* start;
+		C2Di_Glyph* end;
+		float wrapXOffset;
+		u32 newLineNumber;
+	};
+	struct word words[text->words];
+	for (u32 i = 0; i < text->words; i++)
 	{
-		float glyphW = scaleX*cur->width;
-		float glyphX = x+scaleX*cur->xPos;
-		float glyphY = y+dispY*cur->lineNo;
-
-		C2Di_SetTex(cur->sheet);
-		C2Di_Update();
-		C2Di_AppendVtx(glyphX,        glyphY,        glyphZ, cur->texcoord.left,  cur->texcoord.top,    0.0f, 1.0f, color);
-		C2Di_AppendVtx(glyphX,        glyphY+glyphH, glyphZ, cur->texcoord.left,  cur->texcoord.bottom, 0.0f, 1.0f, color);
-		C2Di_AppendVtx(glyphX+glyphW, glyphY,        glyphZ, cur->texcoord.right, cur->texcoord.top,    0.0f, 1.0f, color);
-		C2Di_AppendVtx(glyphX+glyphW, glyphY,        glyphZ, cur->texcoord.right, cur->texcoord.top,    0.0f, 1.0f, color);
-		C2Di_AppendVtx(glyphX,        glyphY+glyphH, glyphZ, cur->texcoord.left,  cur->texcoord.bottom, 0.0f, 1.0f, color);
-		C2Di_AppendVtx(glyphX+glyphW, glyphY+glyphH, glyphZ, cur->texcoord.right, cur->texcoord.bottom, 0.0f, 1.0f, color);
+		words[i].start = NULL;
+		words[i].end = NULL;
+		words[i].wrapXOffset = 0;
+		words[i].newLineNumber = 0;
+	}
+	for (cur = begin; cur != end; cur++)
```

Comment:
Move this code somewhere else so that it doesn't run for the trivial case that doesn't need these calculations?